src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -2989,11 +2989,21 @@ internal static bool CheckValEscape(SyntaxNode node, BoundExpression expr, uint
             {
                 case BoundKind.DefaultLiteral:
                 case BoundKind.DefaultExpression:
-                case BoundKind.Parameter:
                 case BoundKind.ThisReference:
                     // always returnable
                     return true;
 
+                case BoundKind.Parameter:
+                    var parameter = ((BoundParameter)expr).ParameterSymbol;
+                    // params Span<T> value cannot be returned since that would
+                    // prevent sharing repeated allocations at the call-site.
+                    if (parameter.IsParams && !parameter.Type.IsSZArray())
```

Comment:
>IsParams [](http://example.com/codeflow?start=34&length=8)

If I remember correctly, it used to be that this property on its own wasn't sufficient for a parameter to be treated as a "real" `params` parameter. The type and position was important as well. Also, even if parameter is a "real" params, it doesn't mean that `Span<T>` was created by compiler, a user-created span could be passed. Another thing to consider is the fact, that removing `params` is not a binary breaking change, but it looks like it is going to disable the safety check. Perhaps, it would be better to move the safety check to the consumer, where we know exactly what is going on rather than doing this at the definition site?

Reply:
Updated check for `params` and added a PROTOTYPE comment to move the error reporting to the caller.